### PR TITLE
Add more comprehensive testing via tox

### PR DESCRIPTION
**Pull request by pawroman**
Add a tox config for easy testing across multiple Python versions and environments.

Travis should be picking it up.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,14 +2,28 @@ language: python
...
-  - 3.6-dev
+  - 3.7-dev
+
+env:
+  - NUMPY=1.13
+  - NUMPY=1.14
+
+matrix:
+  allow_failures:
+    - python: "3.7-dev"
```
Comment (pawroman) [inline .travis.yml:17]: Currently failing due to our dependencies (numba) not supporting it. Can take it out of allowed failures once all our dependencies support 3.7

File: setup.py [hunk 2]
```diff
@@ -92,17 +99,11 @@ def read_utf8(filename):
...
 
 
-# CI-specific test utilities, e.g. travis, appveyor
-setup_kwargs['extras_require']['ci_test'] = (
-    setup_kwargs['tests_require']
-    + [
-        'codecov',
-        'httpie',
-    ]
-)
```
Comment (pawroman) [inline setup.py:102]: I thought that we should not keep CI-specific dependencies in setup.py, rather move them to `.travis.yml` and `.appveyor.yml`

File: .travis.yml (hunk 1 above)
Comment (dwillmer) [reply .travis.yml:17]: yea, agreed - nice one, thanks

**Comment by dwillmer** [comment]
I'm actually not sold on `tox` at all - currently if i want to run the tests, i run `pytest`, if I want to change the environment and run tests, i run `source activate py35`, and then `pytest` (what i'll actually do is have multiple command windows open with the currently supported environments). Either way, it's not something that slows down development.

More importantly, we currently use github PRs to find out if all the permutations of libraries are working, but tox doesn't test Windows, so even if we merge this PR and swallow the pill with tox/pyenv, we still need to use github PRs on top in order to test all environments.

As a result, I only see tox as being something that adds dependencies and adds overheads (needing to know tox.ini peculiarities as well as pytest.ini peculiarities), but doesn't improve the workflow.

On top of tox, there's no way we can prefer `pyenv` over `conda` - `pyenv` is fine, but the scientific python stack is really much better with conda, and most people we know use conda, so i'd rather stick with that.

I am very happy to be proven wrong 😄 , so it would be great if there's some benefits that i'm not aware of (please let me know), but at the moment tox seems unnecessary to me.

**Comment by dwillmer** [comment]
Closing for now - happy to re-open if the `tox` argument becomes more compelling.

**Pull request closed without merging**